File: swift/Sources/GildedRose/ItemCategory/AgedBrie.swift
//
//  AgedBrie.swift
//  
//
//  Created by Frank Thamel on 2022-08-19.
//

import Foundation

class AgedBrie: Item {
    
    /**
     Override `updateQuality` to provide  the correct implementation for the `AgedBrie` type.
     */
    override func updateQuality() {
        
        // Lowers sellIn value by 1 each day
        sellIn -= 1
        
        // Maximum quality for an item is 50
        guard quality < 50 else { return }
        
        // "Aged Brie" increases in Quality the older it gets
        quality += 1
        
        // Maximum quality for an item is 50
        guard quality < 50 else { return }
        
        // Quality increases twice as fast when passed SellIn date
        guard sellIn < 0 else { return }
        
        quality += 1
        
    }
}

File: ruby/update_handlers/aged_brie.rb
# frozen_string_literal: true

require_relative 'base'

module UpdateHandlers
  class AgedBrie < Base
    private

    def update_sell_in
      item.sell_in -= 1
    end

    def update_quality
      quality_value = quality + 1

      item.quality = normalise(quality_value)
    end
  end
end
